website/project/views/comment.py

```diff
@@ -17,30 +18,69 @@
+                update_comment_node(guid, source_node, destination_node)
+
+            if not (source['provider'] == destination['provider'] and source['provider'] == 'osfstorage'):
+                if not source['path'].endswith('/'):
+                    data = dict(destination)
+                    new_file = FileNode.resolve_class(destination['provider'], FileNode.FILE).get_or_create(destination_node, destination['path'])
+                    if destination['provider'] != 'osfstorage':
+                        new_file.update(revision=None, data=data)
+                else:
+                    new_file = create_file_from_metadata(destination.get('children', []), source, destination, destination_node, obj)
+                    if not new_file:
+                        if source['provider'] == 'box':
+                            new_path = obj.referent.path
+                        else:
+                            new_path = obj.referent.materialized_path.replace(source['materialized'], destination['materialized'])
+                        new_file = FileNode.resolve_class(destination['provider'], FileNode.FILE).get_or_create(destination_node, new_path)
+                        new_file.name = obj.referent.name
+                        new_file.materialized_path = new_path
+                        new_file.save()
+
+                obj.referent = new_file
+                obj.save()
+
+
+def create_file_from_metadata(children, source, destination, destination_node, obj):
     for item in children:
-        if not item.is_file:
-            update_folder_contents(item.children, source_node, destination_node)
-        else:
-            Comment.update(Q('root_target', 'eq', item._id), data={'node': destination_node})
-            # update node record of commented files
-            if item._id in source_node.commented_files:
-                destination_node.commented_files[item._id] = source_node.commented_files[item._id]
-                del source_node.commented_files[item._id]
-                source_node.save()
-                destination_node.save()
+        if item['kind'] == 'folder':
+            return create_file_from_metadata(item.get('children', []), source, destination, destination_node, obj)
+        elif item['kind'] == 'file' and item['materialized'].replace(destination['materialized'], source['materialized']) == obj.referent.materialized_path:
+            data = dict(item)
+            new_file = FileNode.resolve_class(destination['provider'], FileNode.FILE).get_or_create(destination_node, item['path'])
+            if destination['provider'] != 'osfstorage':
+                new_file.update(revision=None, data=data)
+            return new_file
+
+
+def update_comment_node(root_target_id, source_node, destination_node):
+    Comment.update(Q('root_target', 'eq', root_target_id), data={'node': destination_node})
+    destination_node.commented_files[root_target_id] = source_node.commented_files[root_target_id]
```

Comment:
As discussed on flowdock and with @sloria, `Node.commented_files` can be removed for now since it's used when calculating the number of unread file comments, but that unread count is currently not being used and is expensive to calculate.